### PR TITLE
Config Via Parameter Store

### PR DESCRIPTION
**Pull request by QuarpT**
## What does this change?

Frontend configuration - no longer load from S3. Instead loads from parameter store

Alternative to https://github.com/guardian/frontend/pull/18423 after @TBonnin's comments on property versioning.

Config file properties can now point to properties stored in parameter store.

See PR `docs/03-dev-howtos/12-Update-configuration.md`

## Why do configuration this way?

- GDPR requirements 
https://docs.google.com/a/guardian.co.uk/document/d/1OYH08KssBgD3eU_ytwxl5-C6aD0ACyVtzXZEZnxNqT8/edit?usp=sharing. Individual apps only load the configuration they require dependent on the stage. Editing config does not include downloading all production secrets locally and resubmitting them to S3.
- We want configuration properties to be immutable and dependent on git commits / releases.
This allows us to rollback configuration properties via deployments.
- Parameter store performance. It takes 10-15ms on average to load a single property (even with batching). Therefore we only want to download the specific version of a property that an app requires.
- Local overrides. We don't want to apply versioning within `.conf` property names since this will 
interfere with local overrides.
- Potential for storing non-sensitive, stage-specific configuration locally within the frontend repo.

## Further improvements

I think the overrides should go in the local app conf directories.

Then set `-Dconfig.resource` to the overriding `{stage}.conf`. `{stage}.conf` would then begin with an `include` for the shared `global/{stage}.conf` and the app's local `application.conf`.

This would be more in line with conventional Play app config.

## Tested in CODE?

Needs thorough testing

### CODE REVIEW

File: app-config/shared/global/dev.conf [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+assets.path = "param-store@/frontend/dev/assets.path.v1"
+images.path = "param-store@/frontend/dev/images.path.v1"
+static.path = "param-store@/frontend/dev/static.path.v1"
+static.securePath = "param-store@/frontend/dev/static.securePath.v1"
+staticSport.path = "param-store@/frontend/dev/staticSport.path.v1"
+switches.key = "param-store@/frontend/dev/switches.key.v1"
+front.config = "param-store@/frontend/dev/front.config.v1"
+frontend.store = "param-store@/frontend/dev/frontend.store.v1"
+
+# uncomment to test cors locally
```
Comment (TBonnin) [inline app-config/shared/global/dev.conf]: no need I think

File: app-config/shared/global/dev.conf [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+# Properties needed to run Next-Gen
+# these properties should be put inside [HOME_DIRECTORY]/.gu/frontend.properties
+#
+# You should only share this file with other team members, some of the keys in here are not for wider use
+
+############################################################################
+# REQUIRED properties - you need all of these for the project to work
+# you should be fine using the defaults below
+############################################################################
```
Comment (TBonnin) [inline app-config/shared/global/dev.conf]: The 10 lines above are no needed I think

File: app-config/shared/global/dev.conf [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+weather.api.key = "param-store@/frontend/dev/weather.api.key.v1"
+
+riffraff.apikey = "param-store@/frontend/dev/riffraff.apikey.v1"
+riffraff.url = "param-store@/frontend/dev/riffraff.url.v1"
+
+elk.kibana.url = "param-store@/frontend/dev/elk.kibana.url.v1"
+
+############################################################################
+# OPTIONAL properties - you can probably run without these
+############################################################################
```
Comment (TBonnin) [inline app-config/shared/global/dev.conf]: needed comment?

File: common/test/concurrent/FutureSemaphoreTest.scala [hunk 4]
```diff
@@ -11,7 +11,7 @@ class FutureSemaphoreTest extends FlatSpec with Matchers with ScalaFutures {
   def executeSomeTasks[T](futureSemaphore: FutureSemaphore, numberToExecute: Int)(task: => Future[T]): Future[Seq[Try[T]]] = {
     Future.traverse(1 to numberToExecute)(_ =>
       futureSemaphore.execute[Try[T]](
-        Future(Thread.sleep(2)).flatMap(_ => task).map(Success(_))
+        Future(Thread.sleep(10)).flatMap(_ => task).map(Success(_))
```
Comment (TBonnin) [inline common/test/concurrent/FutureSemaphoreTest.scala]: This is not related to the config change right?

**Review by TBonnin — APPROVED** [review]
lgtm. 
⚠️ Make sure nobody has modified the config file in S3 since the last time you uploaded the config into param-store 

I think it would be a good idea to create an admin page (ex: https://frontend.gutools.co.uk/config) to help and facilitate the modification of config by guiding any developers through the 3 required steps:
1. Direct Link to the config files folder in the repo so they are easy to find.
2. Search for parameter-store keys related to a given config setting (search feature of param-store is prefix only, no fuzzy search)
3. Direct link to param-store

File: app-config/shared/global/dev.conf (hunk 3 above)
Comment (QuarpT) [reply app-config/shared/global/dev.conf]: Copied from S3 file

File: common/test/concurrent/FutureSemaphoreTest.scala (hunk 4 above)
Comment (QuarpT) [reply common/test/concurrent/FutureSemaphoreTest.scala]: No, it's already in master too I think

File: app-config/shared/global/dev.conf (hunk 2 above)
Comment (QuarpT) [reply app-config/shared/global/dev.conf]: 👍  This comment was in the S3 config

**Comment by QuarpT** [comment]
@TBonnin I will create a new admin page for config in a separate PR

**Review by sihil — COMMENTED** [review]
My first instinct with this is that there is a huge amount of boilerplate instead of some sane decisions about how keys should be merged together. I feel like keys with `param-store@` should be completely unnecessary as it should be read in and merged implicitly.

Feels like there is a lot of wiring to do for every new config value.

#18423 seemed pretty succinct in comparison and I don't quite follow the switch to this.

**Comment by QuarpT** [comment]
@sihil I also prefer the solution in https://github.com/guardian/frontend/pull/18423

However, if we want to be able to map SCM to parameter store property versions then this PR is the way forward. It's necessary to do `param-store@{name}.{version}` because of parameter store performance. It takes 10-15ms to download a single property (even with batching). We can only afford to load the specific versions of a property the app requires.

The question is, do we want to map SCM to parameter store given the complexity? Speaking to @TBonnin and @nicl the yesterday, the answer seems to be yes.

**Comment by rich-nguyen** [comment]
My vote goes with #18423 for the same reasons as @sihil

**Comment by LATaylor-guardian** [comment]
> My vote goes with #18423 for the same reasons as @sihil
Ditto.

**Comment by TBonnin** [comment]
I like the simplicity of #18423 
However it has 2 flaws in my opinion:
1. Config settings are mutable
2. Modifying config settings is not tied to a deploy

Successful modifications of the config will then be based on the hope no deploy or autoscaling events happen mid changes. In the context of frontend where deployments are numerous and comes from different teams, often with very little coordination I would personally prefer to reduce as much as possible the likelihood . 

Also, requiring a code change when config is being modified is indeed adding a step in the process but this step ensures a better tracability (who change what and when) as well as forcing whoever making the change to carry the modification all the way to production.

It is true that it makes the code more complex. I am happy to accept the price of a little more complexity in this PR is considered too high though and I am fine with going with the popular opinion and merge #18423

**Comment by QuarpT** [comment]
I will go ahead with merging #18423 for now
Thanks for all the feedback!
